### ISSUE TITLE
Dependency on --pid=host

### ISSUE THREAD

**Issue by font** [ceph/ceph-docker#481]
Initial attempts to remove `--pid=host` from the Ceph OSDs resulted in `systemd` errors as a result of #479, which should be resolved with either #478 or #480.

After #479 was resolved, removing `--pid=host` resulted in Ceph OSD and host networking issues. This might be due to multiple Ceph OSD processes in their own container PID namespaces attempting to listen to the same public and cluster network IP address and port. Simple ping tests from either the host or the container to the other Ceph OSD hosts would result in occasional errors and dropped packets.

Here is an attempt to contact `ceph-osd1` and `ceph-osd2` from `ceph-osd0` on their public and cluster network IP addresses:

```
-bash-4.2# hostname
ceph-osd0
-bash-4.2# ping 192.168.125.101
PING 192.168.125.101 (192.168.125.101) 56(84) bytes of data.
ping: sendmsg: Operation not permitted
ping: sendmsg: Operation not permitted
ping: sendmsg: Operation not permitted
ping: sendmsg: Operation not permitted
ping: sendmsg: Operation not permitted
ping: sendmsg: Operation not permitted
64 bytes from 192.168.125.101: icmp_seq=7 ttl=64 time=0.451 ms
64 bytes from 192.168.125.101: icmp_seq=8 ttl=64 time=0.400 ms
64 bytes from 192.168.125.101: icmp_seq=9 ttl=64 time=0.345 ms
64 bytes from 192.168.125.101: icmp_seq=10 ttl=64 time=0.395 ms
64 bytes from 192.168.125.101: icmp_seq=11 ttl=64 time=0.392 ms
64 bytes from 192.168.125.101: icmp_seq=12 ttl=64 time=0.327 ms
64 bytes from 192.168.125.101: icmp_seq=13 ttl=64 time=0.300 ms
64 bytes from 192.168.125.101: icmp_seq=14 ttl=64 time=0.332 ms
^C
--- 192.168.125.101 ping statistics ---
14 packets transmitted, 8 received, 42% packet loss, time 12999ms
rtt min/avg/max/mdev = 0.300/0.367/0.451/0.052 ms
-bash-4.2# ping 192.168.125.102
PING 192.168.125.102 (192.168.125.102) 56(84) bytes of data.
64 bytes from 192.168.125.102: icmp_seq=1 ttl=64 time=0.243 ms
64 bytes from 192.168.125.102: icmp_seq=2 ttl=64 time=0.095 ms
64 bytes from 192.168.125.102: icmp_seq=3 ttl=64 time=0.395 ms
64 bytes from 192.168.125.102: icmp_seq=4 ttl=64 time=0.317 ms
64 bytes from 192.168.125.102: icmp_seq=5 ttl=64 time=0.389 ms
64 bytes from 192.168.125.102: icmp_seq=6 ttl=64 time=0.415 ms
64 bytes from 192.168.125.102: icmp_seq=7 ttl=64 time=0.273 ms
64 bytes from 192.168.125.102: icmp_seq=8 ttl=64 time=0.382 ms
^C
--- 192.168.125.102 ping statistics ---
8 packets transmitted, 8 received, 0% packet loss, time 7000ms
rtt min/avg/max/mdev = 0.095/0.313/0.415/0.102 ms
-bash-4.2# ping 192.168.126.201
PING 192.168.126.201 (192.168.126.201) 56(84) bytes of data.
64 bytes from 192.168.126.201: icmp_seq=8 ttl=64 time=0.357 ms
64 bytes from 192.168.126.201: icmp_seq=9 ttl=64 time=0.151 ms
64 bytes from 192.168.126.201: icmp_seq=10 ttl=64 time=0.394 ms
64 bytes from 192.168.126.201: icmp_seq=11 ttl=64 time=0.392 ms
64 bytes from 192.168.126.201: icmp_seq=12 ttl=64 time=0.395 ms
64 bytes from 192.168.126.201: icmp_seq=13 ttl=64 time=0.463 ms
^C
--- 192.168.126.201 ping statistics ---
13 packets transmitted, 6 received, 53% packet loss, time 12000ms
rtt min/avg/max/mdev = 0.151/0.358/0.463/0.100 ms
-bash-4.2# ping 192.168.126.202
PING 192.168.126.202 (192.168.126.202) 56(84) bytes of data.
ping: sendmsg: Operation not permitted
ping: sendmsg: Operation not permitted
ping: sendmsg: Operation not permitted
ping: sendmsg: Operation not permitted
ping: sendmsg: Operation not permitted
ping: sendmsg: Operation not permitted
ping: sendmsg: Operation not permitted
ping: sendmsg: Operation not permitted
ping: sendmsg: Operation not permitted
64 bytes from 192.168.126.202: icmp_seq=10 ttl=64 time=0.275 ms
64 bytes from 192.168.126.202: icmp_seq=11 ttl=64 time=0.248 ms
64 bytes from 192.168.126.202: icmp_seq=12 ttl=64 time=0.281 ms
64 bytes from 192.168.126.202: icmp_seq=13 ttl=64 time=0.243 ms
64 bytes from 192.168.126.202: icmp_seq=14 ttl=64 time=0.286 ms
^C
--- 192.168.126.202 ping statistics ---
14 packets transmitted, 5 received, 64% packet loss, time 13000ms
rtt min/avg/max/mdev = 0.243/0.266/0.286/0.025 ms
```

This would eventually result in these OSD heartbeat errors as shown below. This example is from the first Ceph OSD device container (`ceph-osd0-osd-devsda`) on the first Ceph OSD host (`ceph-osd0`):

```
+ exec /usr/bin/ceph-osd --cluster ceph -f -i 2 --setuser ceph --setgroup disk
2017-02-08 06:10:41.321127 7f398f6b6a40 -1 osd.2 0 log_to_monitors {default=true}
2017-02-08 06:11:15.324892 7f39862b6700 -1 osd.2 9 heartbeat_check: no reply from 192.168.125.102:6803 osd.3 since back 2017-02-08 06:10:55.123919 front 2017-02-08 06:10:55.123919 (cutoff 2017-02-08 06:10:55.324889)
2017-02-08 06:11:15.324919 7f39862b6700 -1 osd.2 9 heartbeat_check: no reply from 192.168.125.101:6801 osd.4 since back 2017-02-08 06:11:05.425672 front 2017-02-08 06:10:55.123919 (cutoff 2017-02-08 06:10:55.324889)
2017-02-08 06:11:16.325037 7f39862b6700 -1 osd.2 9 heartbeat_check: no reply from 192.168.125.102:6803 osd.3 since back 2017-02-08 06:10:55.123919 front 2017-02-08 06:10:55.123919 (cutoff 2017-02-08 06:10:56.325035)
2017-02-08 06:11:16.325064 7f39862b6700 -1 osd.2 9 heartbeat_check: no reply from 192.168.125.101:6801 osd.4 since back 2017-02-08 06:11:05.425672 front 2017-02-08 06:10:55.123919 (cutoff 2017-02-08 06:10:56.325035)
2017-02-08 06:11:17.325118 7f39862b6700 -1 osd.2 9 heartbeat_check: no reply from 192.168.125.102:6803 osd.3 since back 2017-02-08 06:10:55.123919 front 2017-02-08 06:10:55.123919 (cutoff 2017-02-08 06:10:57.325115)
2017-02-08 06:11:17.325173 7f39862b6700 -1 osd.2 9 heartbeat_check: no reply from 192.168.125.101:6801 osd.4 since back 2017-02-08 06:11:05.425672 front 2017-02-08 06:10:55.123919 (cutoff 2017-02-08 06:10:57.325115)
2017-02-08 06:11:18.325293 7f39862b6700 -1 osd.2 9 heartbeat_check: no reply from 192.168.125.102:6803 osd.3 since back 2017-02-08 06:10:55.123919 front 2017-02-08 06:10:55.123919 (cutoff 2017-02-08 06:10:58.325290)
2017-02-08 06:11:18.325335 7f39862b6700 -1 osd.2 9 heartbeat_check: no reply from 192.168.125.101:6801 osd.4 since back 2017-02-08 06:11:05.425672 front 2017-02-08 06:10:55.123919 (cutoff 2017-02-08 06:10:58.325290)
2017-02-08 06:11:19.325453 7f39862b6700 -1 osd.2 9 heartbeat_check: no reply from 192.168.125.102:6803 osd.3 since back 2017-02-08 06:10:55.123919 front 2017-02-08 06:10:55.123919 (cutoff 2017-02-08 06:10:59.325452)
2017-02-08 06:11:19.325482 7f39862b6700 -1 osd.2 9 heartbeat_check: no reply from 192.168.125.101:6801 osd.4 since back 2017-02-08 06:11:05.425672 front 2017-02-08 06:10:55.123919 (cutoff 2017-02-08 06:10:59.325452)
2017-02-08 06:11:20.325608 7f39862b6700 -1 osd.2 9 heartbeat_check: no reply from 192.168.125.102:6803 osd.3 since back 2017-02-08 06:10:55.123919 front 2017-02-08 06:10:55.123919 (cutoff 2017-02-08 06:11:00.325606)
2017-02-08 06:11:20.325639 7f39862b6700 -1 osd.2 9 heartbeat_check: no reply from 192.168.125.101:6801 osd.4 since back 2017-02-08 06:11:05.425672 front 2017-02-08 06:10:55.123919 (cutoff 2017-02-08 06:11:00.325606)
2017-02-08 06:11:21.325727 7f39862b6700 -1 osd.2 9 heartbeat_check: no reply from 192.168.125.102:6803 osd.3 since back 2017-02-08 06:10:55.123919 front 2017-02-08 06:10:55.123919 (cutoff 2017-02-08 06:11:01.325723)
2017-02-08 06:11:21.325764 7f39862b6700 -1 osd.2 9 heartbeat_check: no reply from 192.168.125.101:6801 osd.4 since back 2017-02-08 06:11:05.425672 front 2017-02-08 06:10:55.123919 (cutoff 2017-02-08 06:11:01.325723)
2017-02-08 06:11:22.325911 7f39862b6700 -1 osd.2 9 heartbeat_check: no reply from 192.168.125.102:6803 osd.3 since back 2017-02-08 06:10:55.123919 front 2017-02-08 06:10:55.123919 (cutoff 2017-02-08 06:11:02.325909)
2017-02-08 06:11:22.325946 7f39862b6700 -1 osd.2 9 heartbeat_check: no reply from 192.168.125.101:6801 osd.4 since back 2017-02-08 06:11:05.425672 front 2017-02-08 06:10:55.123919 (cutoff 2017-02-08 06:11:02.325909)
2017-02-08 06:11:23.326085 7f39862b6700 -1 osd.2 9 heartbeat_check: no reply from 192.168.125.102:6803 osd.3 since back 2017-02-08 06:10:55.123919 front 2017-02-08 06:10:55.123919 (cutoff 2017-02-08 06:11:03.326082)
2017-02-08 06:11:23.326118 7f39862b6700 -1 osd.2 9 heartbeat_check: no reply from 192.168.125.101:6801 osd.4 since back 2017-02-08 06:11:05.425672 front 2017-02-08 06:10:55.123919 (cutoff 2017-02-08 06:11:03.326082)
2017-02-08 06:11:24.326219 7f39862b6700 -1 osd.2 9 heartbeat_check: no reply from 192.168.125.102:6803 osd.3 since back 2017-02-08 06:10:55.123919 front 2017-02-08 06:10:55.123919 (cutoff 2017-02-08 06:11:04.326217)
2017-02-08 06:11:24.326257 7f39862b6700 -1 osd.2 9 heartbeat_check: no reply from 192.168.125.101:6801 osd.4 since back 2017-02-08 06:11:05.425672 front 2017-02-08 06:10:55.123919 (cutoff 2017-02-08 06:11:04.326217)
2017-02-08 06:11:25.326353 7f39862b6700 -1 osd.2 9 heartbeat_check: no reply from 192.168.125.102:6803 osd.3 since back 2017-02-08 06:10:55.123919 front 2017-02-08 06:10:55.123919 (cutoff 2017-02-08 06:11:05.326352)
2017-02-08 06:11:25.326386 7f39862b6700 -1 osd.2 9 heartbeat_check: no reply from 192.168.125.101:6801 osd.4 since back 2017-02-08 06:11:05.425672 front 2017-02-08 06:10:55.123919 (cutoff 2017-02-08 06:11:05.326352)
2017-02-08 06:11:26.326480 7f39862b6700 -1 osd.2 9 heartbeat_check: no reply from 192.168.125.101:6803 osd.1 since back 2017-02-08 06:11:20.928005 front 2017-02-08 06:11:05.425672 (cutoff 2017-02-08 06:11:06.326478)
``` 

A `netstat` in this failure scenario (i.e. without `--pid=host`) results in:

```
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
tcp        0      0 192.168.126.200:6800    0.0.0.0:*               LISTEN      16248/ceph-osd      
tcp        0      0 192.168.125.100:6800    0.0.0.0:*               LISTEN      16282/ceph-osd      
tcp        0      0 192.168.125.100:6801    0.0.0.0:*               LISTEN      16248/ceph-osd      
tcp        0      0 192.168.126.200:6801    0.0.0.0:*               LISTEN      16248/ceph-osd      
tcp        0      0 192.168.125.100:6802    0.0.0.0:*               LISTEN      16248/ceph-osd      
tcp        0      0 192.168.125.100:6804    0.0.0.0:*               LISTEN      16282/ceph-osd      
tcp        0      0 192.168.126.200:6804    0.0.0.0:*               LISTEN      16282/ceph-osd      
tcp        0      0 192.168.126.200:6805    0.0.0.0:*               LISTEN      16282/ceph-osd      
```

Notice how the ports do not increase sequentially from 6800-6803 whereas a good `netstat` when using `--pid=host` should look like this:

```
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 192.168.126.200:6800    0.0.0.0:*               LISTEN      16231/ceph-osd
tcp        0      0 192.168.125.100:6800    0.0.0.0:*               LISTEN      16231/ceph-osd
tcp        0      0 192.168.125.100:6801    0.0.0.0:*               LISTEN      16231/ceph-osd
tcp        0      0 192.168.126.200:6801    0.0.0.0:*               LISTEN      16231/ceph-osd
tcp        0      0 192.168.126.200:6802    0.0.0.0:*               LISTEN      16197/ceph-osd
tcp        0      0 192.168.125.100:6802    0.0.0.0:*               LISTEN      16197/ceph-osd
tcp        0      0 192.168.125.100:6803    0.0.0.0:*               LISTEN      16197/ceph-osd
tcp        0      0 192.168.126.200:6803    0.0.0.0:*               LISTEN      16197/ceph-osd
```

With ports increasing nicely.

These errors seem consistent with what appears to be reported here http://tracker.ceph.com/issues/10763.

To resolve this, Docker now has support for `--pid=container:<name|id>` added in this PR https://github.com/docker/docker/pull/22481. See https://docs.docker.com/engine/reference/run/#/pid-settings---pid for more details.

However, when using the `--pid=container:<name|id>` Docker feature, we need to specify the name of an already running container. This poses a problem when we're bootstrapping multiple Ceph OSDs on the same host using our [templated version of `ceph-osd.service` systemd unit file](https://github.com/ceph/ceph-ansible/blob/1b58afce7fcf0df70ba465d3d25c6c159cda2729/roles/ceph-osd/templates/ceph-osd.service.j2).

For example, say we have 2 devices defined as:

```
devices:
 - /dev/sda
 - /dev/sdb
```

We currently only have 1 systemd service unit file for all Ceph OSD devices on each host, `/dev/sda` and `/dev/sdb` in this case symbolically linked to the 1 templated systemd unit file. But the first device container should not use the `--pid=container:<name|id>` option, while the rest of the device containers should specify the name of the first device container `<name|id>` such as `--pid=container:ceph-osd0-osd-devsda`.

Some ways to solve this come to mind initially:
1. We could create 2 templated systemd service unit files. One for the first container that does not use the `--pid=container:<name|id>` option, and one that does use it and references the first container name. This would require using some Ansible kung fu to link the first device to one unit file, and the rest to the other unit file.
2. We could create a dummy or "debug" Ceph OSD container first, that essentially runs for the entirety of the Ceph OSD cluster, and have 1 systemd service unit file that uses `--pid=container:<name|id>` specifying the name of this dummy/debug container.
3. Remove the templated Jinja2 systemd service unit file and rely on the newer `docker_container` Ansible module along with some Ansible kung fu to use the first device from `ceph_osd_docker_devices` for one task, and the rest of the devices for another task. We may be able to use the same task.
4. Request/implement Docker feature that would allow one to specify `--pid=container:<name|id>` using the name of the container that is about to be created. IOW, making the `--pid` option self-referencing. This would allow the first container to specify its own name in the `--pid=container:<name|id>` option.
5. Keep using `--pid=host`.
6. Others?

**Comment by ktdreyer** [comment]
What if we defined non-overlapping values for `ms_bind_port_min` and `ms_bind_port_max` for each OSD and MDS container? Then they would not interfere with each other?

**Comment by font** [comment]
I'm still not convinced that the port enumeration is the problem. I'm not sure yet how the OSD ports are determined on start-up as I'd have to look at the code, but they definitely do not enumerate the same way with and without `--pid=host`. What I'm certain of is that there appears to be a networking issue that occurs on the host when removing `--pid=host` and the Ceph OSDs come up and attempt to heartbeat each other.

**Comment by font** [comment]
Now that #479 is resolved. I'm closing this issue and tracking it in https://github.com/ceph/ceph-ansible/issues/1297.
